Add engine for side-by-side execution option at command diff (-y)

sami-daniel opened the pull request:
Add -y (or --side-by-side) option to sdiff command to allow viewing of file in side by side mode. This is also the base engine for sdiff command.

renovate[bot] committed:
fix(deps): update rust crate regex to v1.11.1

cakebaker committed:
Merge pull request #102 from uutils/renovate/regex-1.x-lockfile

fix(deps): update rust crate regex to v1.11.1

renovate[bot] committed:
chore(deps): update rust crate tempfile to v3.14.0

cakebaker committed:
Merge pull request #104 from uutils/renovate/tempfile-3.x-lockfile

chore(deps): update rust crate tempfile to v3.14.0

cakebaker committed:
Fix "unused import" warning on Windows

cakebaker committed:
ci: remove CARGO_FEATURES_OPTION

sylvestre committed:
Merge pull request #105 from cakebaker/fix_unused_import_on_windows

Fix "unused import" warning on Windows

oSoMoN committed:
Merge pull request #106 from cakebaker/ci_fix_cargo_features_option

ci: remove `CARGO_FEATURES_OPTION`

renovate[bot] committed:
chore(deps): update codecov/codecov-action action to v5

cakebaker committed:
ci: fix deprecated codecov argument

cakebaker committed:
Merge pull request #107 from uutils/renovate/codecov-codecov-action-5.x

chore(deps): update codecov/codecov-action action to v5

oSoMoN committed:
ci: make sure gpatch is actually being used for tests on MacOS

oSoMoN committed:
Merge pull request #112 from oSoMoN/ci-macos-use-gpatch

ci: make sure gpatch is actually being used for tests on MacOS

renovate[bot] committed:
fix(deps): update rust crate itoa to v1.0.15

oSoMoN committed:
Merge pull request #108 from uutils/renovate/itoa-1.x-lockfile

fix(deps): update rust crate itoa to v1.0.15

renovate[bot] committed:
fix(deps): update rust crate chrono to v0.4.40

oSoMoN committed:
Merge pull request #109 from uutils/renovate/chrono-0.x-lockfile

fix(deps): update rust crate chrono to v0.4.40

renovate[bot] committed:
chore(deps): update rust crate predicates to v3.1.3

oSoMoN committed:
Merge pull request #110 from uutils/renovate/predicates-3.x-lockfile

chore(deps): update rust crate predicates to v3.1.3

renovate[bot] committed:
chore(deps): update rust crate tempfile to v3.17.1

oSoMoN committed:
Merge pull request #111 from uutils/renovate/tempfile-3.x-lockfile

chore(deps): update rust crate tempfile to v3.17.1

renovate[bot] committed:
chore(deps): update rust crate tempfile to v3.18.0

oSoMoN committed:
Merge pull request #113 from uutils/renovate/tempfile-3.x-lockfile

chore(deps): update rust crate tempfile to v3.18.0

renovate[bot] committed:
chore(deps): update rust crate tempfile to v3.19.0

oSoMoN committed:
Merge pull request #114 from uutils/renovate/tempfile-3.x-lockfile

chore(deps): update rust crate tempfile to v3.19.0

renovate[bot] committed:
chore(deps): update rust crate tempfile to v3.19.1

oSoMoN committed:
Merge pull request #115 from uutils/renovate/tempfile-3.x-lockfile

chore(deps): update rust crate tempfile to v3.19.1

renovate[bot] committed:
chore(deps): update rust crate assert_cmd to v2.0.17

oSoMoN committed:
Merge pull request #116 from uutils/renovate/assert_cmd-2.x-lockfile

chore(deps): update rust crate assert_cmd to v2.0.17

sami-daniel committed:
Create side-by-side module and as flag to diff command

sami-daniel committed:
Create limited_string function and tests to it

sami-daniel committed:
Add side_diff module

kov commented on src/side_diff.rs:
```diff
@@ -1,9 +1,408 @@
+            right_ln: &[u8],
+            symbol: &[u8],
+            tab_size: usize,
+        ) {
```
I forgot to update my main branch xD so, in general I think you are complicating the line handling a bit. End of the day all you need is a function that can enforce a width and string slices. I pushed a version of the sdiff patch you had before where I changed it to use diff::slice() and wrote a function to do the limiting, check it out:

https://github.com/kov/diffutils/blob/fixed-sdiff/src/sdiff.rs#L68

As you can see, when you need to print an empty line all you need to do is pass in an empty string slice, &[], for that one.

sami-daniel replied:
All of this was part of the erroneous model that does not represent the reality of GNU's diff -y. Currently these structures have been removed!

kov commented on src/side_diff.rs:
```diff
@@ -1,9 +1,408 @@
+    // }
+
+    let mut already_dispatched = Vec::new();
+    let mut output = Vec::new();
```
You should not need to keep track of what has already been printed. That'll be a lot of memory, and since this is a Vec, every time you print you will be doing a string comparison against *everything* you already printed on the left side, that's quite inefficient.

Also, there is probably no good reason not to just print to stdtout. You can do this instead of accumulating on a Vec:

`let mut output = io::stdout().lock();`

sami-daniel replied:
This is a good reason to remove unnecessary storage from the outputs, but in the future it may complicate the writing of unit tests for the code. Currently, due to the level of complexity, I believe that unit tests are not necessary

kov commented on src/side_diff.rs:
```diff
@@ -1,9 +1,408 @@
+                };
+
+                diff = Diff::new(left_ln, right_ln);
+                dispatch_to_output(&mut output, &diff, &mut already_dispatched);
```
If I understand this logic correctly, the idea is that if you have a line on the right at the same index as the left line you should just print both as if one is being changed into the other, but I don't think that's the correct way to do it. I think what you are trying to do is distinguish between cases in which a line has been added/removed and those in which it has been changed, but that probably requires a bit more work.

I suggest you just make the base case work - you won't get any | symbols, just < > and common text, and then we can study diff's codebase together to understand how it identifies lines that have changed. We will probably need to make a bigger change to how the results are processed for that, as we will likely need to peek at the next result to make that kind of thing work, so a simple for over results won't be enough.

sami-daniel replied:
All of this was part of the erroneous model that does not represent the reality of GNU's diff -y. Currently these structures have been removed and the base case is the only supported now

kov commented on src/side_diff.rs:
```diff
@@ -1,9 +1,408 @@
+        let tab_spaces = 61;
+        let limiter = tab_spaces; // for some reason the str goes only to 61 chars, not 60
+        already_dispatched.push(to_dispatch_val.left_ln.line_ndx);
+        if to_dispatch_val.right_ln.content != vec![] && to_dispatch_val.left_ln.content == vec![] {
```
When checking if an array is empty you should use the .is_empty() method. What you do when you write vec![] is create a new, empty vector, which is wasteful.

kov commented on src/side_diff.rs:
```diff
@@ -1,9 +1,408 @@
+        if to_dispatch_val.right_ln.content != vec![] && to_dispatch_val.left_ln.content == vec![] {
+            push_output(
+                output,
+                "".as_bytes(),
```
You can write literal byte strings the same way as characters, just prepend a b: b""

sami-daniel committed:
Re-exporting side_by_side_diff

sami-daniel committed:
Add must_use to limited_string fn

sami-daniel committed:
Refactor the logic around side_diff mod

kov commented on src/side_diff.rs:
```diff
@@ -0,0 +1,80 @@
+    output.write_all(b" ")?; // {space_char}
+    output.write_all(right_ln)?; // {right_line}
+
+    if cfg!(target_os = "windows") {
+        // {EOL}
+        output.write_all(b"\r")?;
+        output.write_all(b"\n")?;
+    } else {
+        output.write_all(b"\n")?;
+    }
```
You should be able to just do writeln!(output) here, I think?

sami-daniel replied:
Yes, cause its only string, but I guess it doesn't make difference right? Like, the writeln! will evaluate to write_all or write, I think....

sami-daniel replied:
As expected, the `writeln!` expands to a call to the write function of the public implementation, or the `std::io::Write` trait or `std::fmt::Write`. In the end of the day, it will result in practically the same result. I see no apparent reason to use `writeln!` instead of simply writing the bytes in the stream, since we dont apply any formatting to those texts

kov replied:
I mean you can do just writeln!(output); instead of the whole if/else block. No reason to manually handle windows vs unix-like just for newline.

sami-daniel replied:
Now I get it

kov commented on src/side_diff.rs:
```diff
@@ -0,0 +1,80 @@
+
+pub fn diff(from_file: &Buf, to_file: &Buf) -> Buf {
+    //      ^ The left file  ^ The right file
+    fn split_lines(input: &[u8]) -> Vec<&[u8]> {
+        input.split(|&c| c == b'\n').collect()
+    }
```
Since this function is just a simple single line, it's probably better to just have the line in place, no?

sami-daniel replied:
I believe that if a line repeats the exactly same pattern than other, this pattern must be transformed into a function to avoid repetition right?

sami-daniel replied:
It might make sense to remove this function because in every code this same line of code is repeated for both files. Makes sense : )

kov commented on src/utils.rs:
```diff
@@ -99,6 +98,16 @@ pub fn report_failure_to_read_input_file(
     );
 }
 
+/// Limits an string at an certain limiter position. This can broke the
+/// encoding of a specific char where it has been cuted.
```
English typos here, 'a' not 'an' before 'string' and 'certain', 'break' instead of 'broke', 'cut' instead of 'cuted'.

sami-daniel replied:
Thanks : )

kov commented on src/utils.rs:
```diff
@@ -99,6 +98,16 @@ pub fn report_failure_to_read_input_file(
+/// encoding of a specific char where it has been cuted.
+#[must_use]
+pub fn limited_string<'a, T: AsRef<[u8]>>(orig: &'a T, limiter: usize) -> Vec<u8> {
+    // TODO: Verify if we have broke the enconding of the char
```
'have broken' or 'broke' (without have)

sami-daniel replied:
Thanks : )

kov commented on src/side_diff.rs:
```diff
@@ -0,0 +1,80 @@
+            Result::Left(left_ln) => {
+                push_output(
+                    &mut output,
+                    &limited_string(left_ln, tab_size),
```
Do you really need a separate function for this? Feels like a bit of a waste, especially since you're creating a new Vec, and you can just use the slice directly here without copying anything.

Or if you prefer keeping the function to avoid repeating the slicing, you should make it take in the &[u8] you have here and return a slice into it, without copying. Something like:

```rust
pub fn truncate_slice<'a>(orig: &'a [u8], limiter: usize) -> &'a [u8] {
    // TODO: Verify if we have broken the enconding of the char when we cut it.
    &orig[..orig.len().min(limiter)]
}
```

Looking at GNU diff I think this logic will end up all in push_output or something that calls, anyway, as you will have to go byte by byte, to handle all of the special cases.

sami-daniel replied:
I created this function just thinking that will be useful not only to side_by_side diff. Makes sense return a reference to the original slice...

sami-daniel replied:
Also, I think its interesting to create a function just for this because I believe that other tools in the diff set would have their lives made easier if they had a place where they could cut their string without worrying about breaking the encoding haha

kov reviewed (changes requested):
Looking good =) a few more suggestions.

kov commented on src/side_diff.rs:
```diff
@@ -0,0 +1,80 @@
+use diff::Result;
+use std::{io::{stdout, StdoutLock, Write}, vec};
+
+type Buf = Vec<u8>;
```
Oh btw I keep forgetting, I don't think this type alias adds any value. It obscures we're dealing with a simple Vec<u8> and doesn't gain us anything?

sami-daniel replied:
It was just because I don't like typing '\<u8>', lol. I'm going to remove it

sami-daniel committed:
Fix documentation grammar and optimize the ouput of limited_string fn:

- Now, the limited_string function only accepts &[u8] as the original parameter, because we are working with possible long strings and the as_ref call can become a possible cause of a mem leak.

- Now, the return type is a ref to the original slice. This is done cause is unnescessary to create a new vec![]

sami-daniel committed:
Remove the type Buf for avoiding confusion and remove the function split_lines. This was done to improve code readability and follow standards written by other devs

kov commented on src/utils.rs:
```diff
@@ -221,36 +220,36 @@ mod tests {
         #[test]
         fn empty_orig_returns_empty() {
-            let orig = "";
+            let orig = "".as_bytes();
```
use b"" instead

sami-daniel replied:
I forgot this

kov reviewed (commented):
With these two small comments I think I am happy and you got some practice ;D. Time to get a real review and deal with the real maintainer's comments.

sami-daniel commented:
> With these two small comments I think I am happy and you got some practice ;D. Time to get a real review and deal with the real maintainer's comments.

That was an interesting experience, let's see what they think : )
